### PR TITLE
Add Qwen3 Audio Encoder

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

Original author @eitanporat in https://github.com/AI-Hypercomputer/maxtext/pull/2726

Revise according to the original PR comments:
* rename flags for consistency and clean up usage
* make AudioEncoder an NNX module
* remove SinusoidsPositionEmbedding, replaced with expanded existing PositionalEmbedding module
* remove Qwen3OmniAudioModel in qwen3.py since we use the AudioEncoder class in decoder.py instead
* add precision related flags

# Tests
tests/check_qwen3_omni_audio_vs_reference.py all pass

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/encoders.py [hunk 1]
```diff
@@ -76,6 +76,43 @@ def __call__(self, input_images, deterministic=False):
     return embeddings
 
 
+class AudioEncoder(nnx.Module):
```
Comment (hengtaoguo) [inline src/MaxText/layers/encoders.py:79]: Thanks for implementing it as NNX!

File: src/MaxText/layers/decoders.py [hunk 2]
```diff
@@ -581,19 +583,30 @@ def _apply_embedding(
...
-            image_masks=image_masks,
+            token_masks=image_masks,
         )
       # TODO(hengtaoguo): Add support for other multimodal models such as Llama4, refactor if needed
       else:
         raise ValueError(f"Unsupported model_name for multimodal: {cfg.model_name}")
 
+    if audio_embeddings is not None and cfg.use_audio:
+      if cfg.model_name in ["qwen3-omni-30b-a3b"]:
+        y = multimodal_utils.merge_mm_embeddings(
```
Comment (hengtaoguo) [inline src/MaxText/layers/decoders.py:596]: Does it support "video/audio interleaving" at the moment? Or it's in following PRs?

File: src/MaxText/layers/models.py [hunk 3]
```diff
@@ -172,6 +187,8 @@ def __call__(
         bidirectional_mask=bidirectional_mask,
         image_embeddings=image_embeddings,
         image_masks=encoder_image_masks,
+        audio_embeddings=audio_embeddings,
+        audio_masks=audio_masks,
```
Comment (hengtaoguo) [inline src/MaxText/layers/models.py:191]: I feel like we can consolidate all multimodal inputs into a class and pass this single variable to the model. But that can be a follow up later.

File: src/MaxText/layers/qwen3.py [hunk 4]
```diff
@@ -1704,6 +1704,322 @@ def qwen3omni_visionprojector_as_linen(config: Config, mesh: Mesh) -> nn.Module:
...
+    return hidden_states
+
+
+@dataclasses.dataclass(repr=False)
+class Qwen3OmniAudioProjector(nnx.Module):
+  """Projection layer that converts audio encoder output to model embedding space."""
+
+  config: Config
+  proj1: DenseGeneral
+  proj2: DenseGeneral
```
Comment (hengtaoguo) [inline src/MaxText/layers/qwen3.py]: Are these three lines necessary?

File: src/MaxText/layers/qwen3.py [hunk 5]
```diff
@@ -1704,6 +1704,322 @@ def qwen3omni_visionprojector_as_linen(config: Config, mesh: Mesh) -> nn.Module:
...
+        quant=None,  # No quantization
+        model_mode=None,  # Not needed for encoder
+        rngs=rngs,
+    )
+
+  def __call__(
+      self,
+      hidden_states: Array,
+      deterministic: bool = False,
+  ):
```
Comment (hengtaoguo) [inline src/MaxText/layers/qwen3.py:1783]: Can we also add an input/output shape comment here?

File: tests/check_qwen3_omni_audio_vs_reference.py
Comment (hengtaoguo) [inline tests/check_qwen3_omni_audio_vs_reference.py:1]: Would you prefer to create this standalone check, or merge it with https://github.com/AI-Hypercomputer/maxtext/blob/main/tests/check_qwen3_embedding_vs_reference.py?

**Comment by hengtaoguo** [comment]
Thank you for the great work!

File: src/MaxText/layers/models.py (hunk 3 above)
Comment (NicoGrande) [reply src/MaxText/layers/models.py:191]: +1 to this!

**Review by NicoGrande — APPROVED** [review]
LGTM

File: src/MaxText/layers/decoders.py (hunk 2 above)
Comment (aireenmei) [reply src/MaxText/layers/decoders.py:596]: This PR doesn't support. It will be added when we revise and merge this PR https://github.com/AI-Hypercomputer/maxtext/pull/2731

File: src/MaxText/layers/qwen3.py (hunk 4 above)
Comment (aireenmei) [reply src/MaxText/layers/qwen3.py]: Good catch, removed.

File: src/MaxText/layers/models.py (hunk 3 above)
Comment (aireenmei) [reply src/MaxText/layers/models.py:191]: I added TODO for follow up later

File: src/MaxText/layers/qwen3.py (hunk 5 above)
Comment (aireenmei) [reply src/MaxText/layers/qwen3.py:1783]: done

File: tests/check_qwen3_omni_audio_vs_reference.py
Comment (aireenmei) [reply tests/check_qwen3_omni_audio_vs_reference.py:1]: Good point, I merged these two files into tests/unit/qwen3_omni_layers_test.py
